python/lsst/ctrl/mpexec/cmdLineFwk.py

```diff
@@ -66,6 +76,148 @@
 _LOG = logging.getLogger(__name__.partition(".")[2])
 
 
+@dataclass
+class _OutputChainedCollectionInfo:
+
+    def __init__(self, registry: Registry, name: str):
+        self.name = name
+        try:
+            self.chain = list(registry.getCollectionChain(name))
+            self.exists = True
+        except MissingCollectionError:
+            self.chain = []
+            self.exists = False
+
+    def __str__(self):
+        return self.name
+
+    name: str
+    exists: bool
+    chain: List[Tuple[str, DatasetTypeRestriction]]
+
+
+@dataclass
+class _OutputRunCollectionInfo:
+
+    def __init__(self, registry: Registry, name: str):
+        self.name = name
+        try:
+            actualType = registry.getCollectionType(name)
+            if actualType is not CollectionType.RUN:
+                raise TypeError(f"Collection '{name}' exists but has type {actualType.name}, not RUN.")
+            self.exists = True
+        except MissingCollectionError:
+            self.exists = False
+
+    name: str
+    exists: bool
+
+
+@dataclass
+class _ButlerFactory:
+
+    def __init__(self, registry: Registry, args: argparse.Namespace, writeable: bool):
+        if args.output is not None:
+            self.output = _OutputChainedCollectionInfo(registry, args.output)
+        else:
+            self.output = None
+        if args.output_run is not None:
+            self.outputRun = _OutputRunCollectionInfo(registry, args.output_run)
+        elif self.output is not None:
+            if args.extend_run:
+                runName, _ = self.output.chain[0]
+            else:
+                runName = "{}/{:%Y%m%dT%Hh%Mm%Ss}".format(self.output, datetime.datetime.now())
```

Comment:
Happy to remove the hms if you think a version with no time delimiter is better; I was worried it would be unreadable with nothing, but I didn't want to use `:` in a directory name.

It should be unique, but all the scenarios I can imagine in which there might be a collision would be indicative of a higher-level logic problem:
 - if we start parallel jobs that are supposed to write to the same run collection, and they race to create it and get to pick their own timestamps, it's a problem if they end up with _different_ run names - any code doing that should create the run itself first and pass that down to the parallel jobs;
 - if different people start distinct jobs with the same chained collection name, they're either not coordinating a joint processing campaign well or one of them is writing to someone else's namespace, and there should be an error.

I believe the fact that `--extend-run` or `--replace-run` are valid if and only if the run already exists guards against both cases, even if it's only the loser of the race that sees the error.

Reply:
I myself am used to ISO format so I have no problem reading that. Colons are certainly better be avoided in paths. If it is just a random/unique name and no one is supposed to convert it back to timestamp then I guess it does not matter, should be ok to keep `hms`.

For uniqueness my worry was about use case when the same person starts multiple jobs which should write to unique output run collections, that could cause a collision with time-based unique names. I guess for that use case it's reasonable to require that `--output` option be unique?

Reply:
Would milliseconds being included in the date be enough?

Reply:
Femtoseconds would probably work better 🙂

Reply:
If anything I'd be more inclined to round to the nearest 5 or 10 seconds, because making the names more likely to be unique seems more likely to mask higher-level bugs - in other words, I think it's much more likely for a user to start multiple jobs at the same time that they expect to write to the _same_ run (e.g. by launching multiple processes themselves instead of using our tools to launch multiple processes), and it's better for that to fail than for it to write to different runs.  But 5 seconds is probably too long for quick try-fail-repeat debugging work, and it's not like relying on time collisions is a reliable way to avoid that problem anyway.  But I really don't think there is a problem with 1s resolution.